README.md
# calvin-scripts

* `visualize_dataset.py`: script from calvin developers. Try `pip install opencv-python` to make `import cv2` work.

* `calvin_extract_numbers.py,jl`: extracts the numeric fields of calvin data files in current directory and prints them to stdout in tab separated format. The fields are given below. Only file-id(1), actions(7), rel_actions(7), robot_obs(15), scene_obs(24) are printed (54 columns). https://github.com/mees/calvin/blob/main/dataset/README.md has more details about the splits and the fields. We also provide description and statistics in the next section.

* `debug-training.tsv`, `debug-validation.tsv`, etc.: https://github.com/denizyuret/calvin-scripts/releases/tag/v0.0.1 The release page has the output of calvin_extract_numbers scripts.

* `calvin_scene_info.py`: Read and print info in the scene_info.npy, ep_lens.npy, ep_start_end_ids.npy files in each calvin data directory.

* `calvin_extract_unique_annotations.py`: Print all the unique task_ids and language annotations in the dataset. There are 34 unique task ids and 389 unique annotations.

* `zcat D-validation.tsv.gz | python calvin_diffs.py`: prints out the differences with the previous line for all but the first line.

* `zcat D-validation.tsv.gz | python calvin_episodes.py`: Tries to guess the episode boundaries if xyz of successive frames differ by more than 8.5 std.

* `zcat D-validation.tsv.gz | python calvin_intervals.py`: Prints out intervals based on idnum discontinuities.

* `zcat D-validation.tsv.gz | python calvin_summarize_numbers.py`: Print statistics for each column.

* `python visualize_calvin_npz.py /datasets/calvin/D/validation/episode_0000000.npz`: Visualize a single frame from a given npz file.

* `python calvin_visualize.py`: Visualize a series of frames in the current directory, a more detailed version of the original visualize_dataset.py.

* `calvin_controller_coordinates.py`, `calvin_controller_regression.py`, `calvin_controller_xyz.py`, `calvin_controller_check.py`: Scripts I tried to discover, output and check the controller (button etc) coordinates with.

* `calvin_extract_tactile.py`: Extract the mean pixels of `depth_tactile` (2) and `rgb_tactile` images. Saved as e.g. `D-validation-tactile.tsv.gz`.

* `calvin_fix_tactile.py`: The initial version of extract_tactile did not normalize the values, depth pixels were too small, rgb pixels were too large. I fixed these using this script, and named the normalized files e.g. `D-validation-tactile2.tsv.gz`. The extract script is also fixed now and normalizes.

* `python calvin_extract_language.py auto_lang_ann.npy`: extract start, end, task, instruction triples in tab separated format

* `python calvin_scene_diff.py D/validation`: extract and normalize scene coordinate differences, saved in sdiff files.

* `zcat data/ABC(D)-training.tsv.gz | perl calvin_scene_fix_abc(d).pl | gzip > data/ABC(D)fixed-training.tsv.gz`: fix the red-pink and red-blue swaps in scenes A and C respectively.

* `python calvin_convert_npz.py ABCD training`: create an npz file merging all features to one array and attached meta information.

* `calvindataset.py`: CalvinDataset class, reads from npz file and subclasses Dataset. Deprecating loaddata.py.


## calvin-files

* episode_XXXXXXX.npz: Each frame is represented in a file named episode_idnum.npz, consecutive idnums indicate consecutive frames (with the exception of episode transitions I guess). Other files indicating the contents: 
* scene_info.npy indicates the first and last frame numbers for a particular directory as well as which scene (A,B,C,D) they come from (although there seems to be some confusion about this). It only exists in */training but describes the union of training and validation.
* ep_start_end_ids.npy indicates the start and end idnums of segments in that particular directory.
* ep_lens.npy indicates the lengths of segments given by ep_start_end_ids.npy.
* statistics.yaml gives basic stats for numeric variables.

| directory | frames | scene_info.npy |
| --------- | ------ | -------------- |
| debug/training | 2771 | {'calvin_scene_D': [358482, 361252]} |
| debug/validation | 1675 | {'calvin_scene_D': [553567, 555241]} |
| D/training | 512077 | {'calvin_scene_D': [0, 611098]} |
| D/validation | 99022 | . |
| ABC/training | 1795045 | {'calvin_scene_B': [0, 598909], 'calvin_scene_C': [598910, 1191338], 'calvin_scene_A': [1191339, 1795044]} |
| ABC/validation | 99022 | . |
| ABCD/training | 2307126 | {'calvin_scene_D': [0, 611098], 'calvin_scene_B': [611099, 1210008], 'calvin_scene_C': [1210009, 1802437], 'calvin_scene_A': [1802438, 2406143]} |
| ABCD/validation | 99022 | . |

The validation directories for ABCD, ABC, and D have the same content: calvin_scene_D:
0:53818, 219635:244284, 399946:420498. (99022 frames). However note that the language
annotations in the D split is different from the ones in the ABC/ABCD splits.

The training directory for D has the rest of the scene D data: calvin_scene_D: 53819:219634,
244285:399945, 420499:611098. (512077 frames).  (scene_info.npy says scene_A but should be
scene_D).

The training directory for ABCD starts with the same content as D/training (except for
several off-by-one errors) followed by B (598910), C (592429), and A (603706) sections which
are identical to ABC/training but renumbered.

The scenes differ by desk color and drawer positioning. The objects look the same.


## calvin-data

A summary of all data (including images) in episode_XXXXXXX.npz files (each represents a single 1/30 sec frame):

```
# julia> for f in data[:files]; println(f, "\t", summary(get(data, f))); end
# actions	7-element Vector{Float64}
# rel_actions	7-element Vector{Float64}
# robot_obs	15-element Vector{Float64}
# scene_obs	24-element Vector{Float64}
# rgb_static	200×200×3 Array{UInt8, 3}
# rgb_gripper	84×84×3 Array{UInt8, 3}
# rgb_tactile	160×120×6 Array{UInt8, 3}
# depth_static	200×200 Matrix{Float32}
# depth_gripper	84×84 Matrix{Float32}
# depth_tactile	160×120×2 Array{Float32, 3}
```

## calvin-tsv-fields

The fields in the output of calvin_extract_numbers.py (files like D-training.tsv.gz) is as follows:
(The npz files keep idnum separate, so column 0 is actx and in general subtract 1 from the following to get column index).
(The ranges given below are from D-validation).

00. idnum
01. actions/x [-0.42,0.37] (tcp (tool center point) position: x,y,z in absolute world coordinates in meters: we want in the next frame, i.e. act[t] = tcp[t+1])
02. actions/y [-0.46,0.10]
03. actions/z [0.30,0.69]
04. actions/a [-pi,pi] (tcp orientation (3): euler angles a,b,c in absolute world coordinates)
05. actions/b [-0.46,0.32]
06. actions/c [-pi,pi]
07. actions/g [-1 or 1] (gripper_action (1): binary close=-1, open=1)
08. rel_actions/x [-1,1] (tcp position (3): x,y,z in relative world coordinates normalized and clipped to (-1, 1) with scaling factor 50: rel[t]=normalize(act[t]-tcp[t]))
09. rel_actions/y [-1,1] (normalize_dist=clip(act-obs,-0.02,0.02)/0.02, normalized_angle=clip(((act-obs) + pi) % (2*pi) - pi, -0.05, 0.05)/0.05
10. rel_actions/z [-1,1]
11. rel_actions/a [-1,1] (tcp orientation (3): euler angles a,b,c in relative world coordinates normalized and clipped to (-1, 1) with scaling factor 20)
12. rel_actions/b [-1,1]
13. rel_actions/c [-1,1]
14. rel_actions/g [-1 or 1] (gripper_action (1): binary close=-1, open=1)
15. robot_obs/x [-0.42,0.37] (tcp position (3): x,y,z in world coordinates: current position of tcp, i.e. tcp[t]=act[t-1])
16. robot_obs/y [-0.46,0.10]
17. robot_obs/z [0.30,0.69]
18. robot_obs/a [-pi,pi] (tcp orientation (3): euler angles a,b,c in world coordinates)
19. robot_obs/b [-0.46,0.32]
20. robot_obs/c [-pi,pi]
21. robot_obs/w [-0.02,0.08] (gripper opening width (1): in meters)
22. robot_obs/j1 [-1.67,0.41] (arm_joint_states (7): in rad)
23. robot_obs/j2 [-0.11,1.44]
24. robot_obs/j3 [1.21,2.55]
25. robot_obs/j4 [-2.80,-0.50]
26. robot_obs/j5 [-1.65,0.17]
27. robot_obs/j6 [1.27,2.56]
28. robot_obs/j7 [-1.20,2.46]
29. robot_obs/g  [-1 or 1] (gripper_action (1): binary close = -1, open = 1)
30. scene_obs/sliding_door [-0.002,0.28] (relative coordinates in meters)
31. scene_obs/drawer [0.0,0.22] (relative coordinates in meters)
32. scene_obs/button [0.0,0.034] (relative coordinates, scaled)
33. scene_obs/switch [0.04,0.09] (relative coordinates, scaled)
34. scene_obs/lightbulb [0 or 1] (1): on=1, off=0
35. scene_obs/green light [0 or 1] (1): on=1, off=0
36. scene_obs/redx (red block (6): (x, y, z, euler_x, euler_y, euler_z)
37. scene_obs/redy
38. scene_obs/redz
39. scene_obs/reda
40. scene_obs/redb
41. scene_obs/redc
42. scene_obs/bluex (blue block (6): (x, y, z, euler_x, euler_y, euler_z)
43. scene_obs/bluey
44. scene_obs/bluez
45. scene_obs/bluea
46. scene_obs/blueb
47. scene_obs/bluec
48. scene_obs/pinkx (pink block (6): (x, y, z, euler_x, euler_y, euler_z)
49. scene_obs/pinky
50. scene_obs/pinkz
51. scene_obs/pinka
52. scene_obs/pinkb
53. scene_obs/pinkc

## calvin-controller-coordinates

The fields in the output of calvin_controller_xyz.py (files like D-training-controllers.tsv.gz) are as follows:
The coordinates give the location tcp should be at to move the controller at each point in time.
For details of the coordinate calculation see calvin_controller_xyz.md and calvin_controller_xyz.py.

00.00 idnum
01.54 slider.x
02.55 slider.y
03.56 slider.z
04.57 drawer.x
05.58 drawer.y
06.59 drawer.z
07.60 button.x
08.61 button.y
09.62 button.z
10.63 switch.x
11.64 switch.y
12.65 switch.z

## calvin-tactile-coordinates

The fields in the output of the calvin_extract_tactile.py script (files like `D-validation-tactile2.tsv.gz`). These give the average pixel of the tactile images. The depth pixels were normalized with x100, the rgb pixels were normalized with /255.0.

00.00 idnum
01.66 depth_tactile1
02.67 depth_tactile2
03.68 rgb_tactile1_r
04.69 rgb_tactile1_g
05.70 rgb_tactile1_b
06.71 rgb_tactile2_r
07.72 rgb_tactile2_g
08.73 rgb_tactile2_b

## calvin-scene-changes

These are the normalized differences in scene coordinates: normalize(x[next]-x[curr]).

Normalization similar to rel_actions based on https://github.com/mees/calvin_env/blob/797142c588c21e76717268b7b430958dbd13bf48/calvin_env/utils/utils.py#L160

00.00 idnum
01.74 scene_obs/sliding_door
02.75 scene_obs/drawer
03.76 scene_obs/button
04.77 scene_obs/switch
05.78 scene_obs/lightbulb
06.79 scene_obs/green_light
07.80 scene_obs/redx
08.81 scene_obs/redy
09.82 scene_obs/redz
10.83 scene_obs/reda
11.84 scene_obs/redb
12.85 scene_obs/redc
13.86 scene_obs/bluex
14.87 scene_obs/bluey
15.88 scene_obs/bluez
16.89 scene_obs/bluea
17.90 scene_obs/blueb
18.91 scene_obs/bluec
19.92 scene_obs/pinkx
20.93 scene_obs/pinky
21.94 scene_obs/pinkz
22.95 scene_obs/pinka
23.96 scene_obs/pinkb
24.97 scene_obs/pinkc


## calvin-annotation-statistics

Each language annotation file is class (task) balanced, roughly equal number of each class
for each dataset.  ABC-validation and ABCD-validation are buggy and should not be used: they
padded the annotation numbers by repeating the same annotations multiple times. Use
D-validation instead.

| dataset | % frames annotated | frames | annots | contig tasks | frames/annot | % frames multiply annotated |
| --------| ------------------ | ------ | ------ | ------------ | ------------ | --------------------------- |
| ABCD-training    | .3734 | 2307126 | 22966 | 14176 | 100.46 | .3650 |
| ABC-training     | .3760 | 1795045 | 17870 | 11102 | 100.45 | .3608 |
| D-training       | .3813 |  512077 |  5124 |  3251 |  99.94 | .3528 |
| debug-training   | .1364 |    2771 |     9 |     7 | 307.89 | .3306 |
| ABCD-validation  | .0634 |   99022 |  1087 |    90 |  91.10 | .9154 |
| ABC-validation   | .0634 |   99022 |  1087 |    90 |  91.10 | .9154 |
| D-validation     | .3660 |   99022 |  1011 |   605 |  97.94 | .3755 |
| debug-validation | .2036 |    1675 |     8 |     6 | 209.38 | .1994 |


ChangeLog
2023-08-18  dyuret  <dyuret@login02.kuacc.ku.edu.tr>

	* calvin_ssl_baselines.py: Measures prediction accuracy of simple baselines for tcp prediction.
	- We need to treat angles, distances and binary features appropriately.
	- Using squared error on features with different units is bad.
	- xyz features have units of meters.
	- abc features have units of radians.
	- controller features are scaled (we can use the derived xyz versions, e.g. slider.x).
	- rel features are scaled (xyz have units of 2cm, abc have units of 0.05 radians).
	- actg, relg, armg are -1/+1 binary.
	- lightbulb, greenlight are 0/1 binary.

	* input-vs-output:
	- The input scaling does not effect the loss, as long as we keep the average value around 1, sgd should work fine.
	- The output scaling directly effects the relative contributions of different outputs to the loss, so be more careful there.

	* meters-vs-radians:
	- In order to combine meters vs radians in loss, we need to bring them to similar units.
	- Because of the wrap-around problem with radians, we already convert them to (sin,cos) pairs.
	- (sin,cos) are in fact xy coordinates on a unit circle.
	- The gripper width is 8cm, we can take 10cm to be the diameter (5cm radius) of our "unit" circle.
	- To work in cm, we'd have to scale xyz by *100 and sincos(abc) with *5.
	- In the rel transform calvin scales xyz by *50 and radians by *20.
	- Our 5cm unit circle has circumference of 31.4cm corresponding to 2pi (6.28) radians. The radian->cm scaling factor would be 5.
	- So, instead of rel coordinates *50/*20 scaling we should use *100/*5 scaling and use sincos instead of radians.


2023-06-03  dyuret  <dyuret@login03.kuacc.ku.edu.tr>

	* eval:
	emrecanacikgoz/robot-language clone main branch
	cd evaluation
	read README
	downgrade numpy (create calvin-eval env)
	train-bc.py: training script for behavior cloning
	policies/behavior_cloning.py: model definition
	lightning model needs a step method step(self, state, task_id) => action_pred[7D] xyz, abc, gripper[-1,+1]
	experiments/ => model gets saved
	cd evaluation/ => do the evaluation here
	model file is copied here
	evaluate_multistep.py
	evaluate_singlestep.py: we are using this now
	needs TRAIN_FOLDER: simulator config file is here
	DATASET_PATH: /datasets/calvin/D/
	CONF_DIR: points to calvin directory
	DEBUG=True: saves videos
	ENV=None: use their environment, not our own.
	checkpoint: use your model
	import model.py file in the beginning
	model = Transformer.load_from_checkpoint()
	model.cuda: we need to do it?
	evaluation/utils.py: TASK_TO_ID_DICT: frame-ids in D that corresponds to particular task.
	back in evaluate_singlestep.py:
	rollout function is in rollout.py.
	rollout.py:29 determines input to model: episode[]
	env.step() executes one step in simulator
	action comes from model
	env.step gives obs, current_info
	get_task_info_for_set: returns true/false for success
	original code: evaluate_policy_singlestep.py - bug-free rollout.

2023-03-29  dyuret  <dyuret@login03.kuacc.ku.edu.tr>

	* warning:
	/userfiles/dyuret/.julia/conda/3/x86_64/lib/python3.10/site-packages/torch/nn/modules/transformer.py:544: UserWarning: Converting mask without torch.bool dtype to bool; this will negatively affect performance. Prefer to use a boolean mask directly. (Triggered internally at /opt/conda/conda-bld/pytorch_1678411187366/work/aten/src/ATen/native/transformers/attention.cpp:150.)

	* gpt.py: Experiments on the D dataset.

	### robot: state,action=>next-state ###
	tcp+0,act+0=>tcp+1: 4921796: min rmsd=0.0422, converges to < 0.05 in 500 steps. {'batch_size': 32, 'max_steps': 100000, 'hidden_size': 512, 'num_layers': 2, 'num_heads': 2, 'dropout': 0, 'weight_decay': 0, 'learning_rate': 0.0001, 'data': 'D'}{'in_features': [('tcp', 0), ('act', 0)], 'out_features': [('tcp', 1)], 'window': 1024}
	  here min-rmsd for xyz = 0.0009, for abc=0.06
	tcp+0,rel+0=>tcp+1: 4921799: min rmsd=0.4470 with simple scaling, converges to 0.35 in 2000 steps. {'batch_size': 32, 'max_steps': 100000, 'hidden_size': 512, 'num_layers': 2, 'num_heads': 2, 'dropout': 0, 'weight_decay': 0, 'learning_rate': 0.0001, 'data': 'D'}{'in_features': [('tcp', 0), ('rel', 0)], 'out_features': [('tcp', 1)], 'window': 1024}
	  here min-rmsd for xyz = 0.0009, for abc=0.6321: we are definitely suffering from angle conversion, **try sin/cos?** try xyz only?
	tcp3+0,act3+0=>tcp3+1: 4921800: min rmsd=0.0009, converges to 0.02286 in 300 steps, 0.00375 in 7000. {'batch_size': 32, 'max_steps': 100000, 'hidden_size': 512, 'num_layers': 2, 'num_heads': 2, 'dropout': 0, 'weight_decay': 0, 'learning_rate': 0.0001, 'data': 'D'}{'in_features': [('tcp3', 0), ('act3', 0)], 'out_features': [('tcp3', 1)], 'window': 1024}
	  here converged to larger loss than min-rmsd: try hparams?
	tcp3+0,rel3+0=>tcp3+1: 4921801: min rmsd=0.0009, converges to 0.006 in 5000 steps.  {'batch_size': 32, 'max_steps': 100000, 'hidden_size': 512, 'num_layers': 2, 'num_heads': 2, 'dropout': 0, 'weight_decay': 0, 'learning_rate': 0.0001, 'data': 'D'}{'in_features': [('tcp3', 0), ('rel3', 0)], 'out_features': [('tcp3', 1)], 'window': 1024}
	  looks good.
	arm+0,act3+0=>arm+1: 4921829: converges to 0.014 in 1800, <0.01 in 4500, 0.00681 in 10K. {'batch_size': 32, 'max_steps': 10000, 'hidden_size': 512, 'num_layers': 2, 'num_heads': 2, 'dropout': 0, 'weight_decay': 0, 'learning_rate': 0.0001, 'data': 'D'}{'in_features': [('arm', 0), ('act3', 0)], 'out_features': [('arm', 1)], 'window': 1024}
	  looks good.
	arm+0,rel3+0=>arm+1: 4921830: converges to 0.013 in 1100, <0.01 in 3700, 0.00621 in 10K. {'batch_size': 32, 'max_steps': 10000, 'hidden_size': 512, 'num_layers': 2, 'num_heads': 2, 'dropout': 0, 'weight_decay': 0, 'learning_rate': 0.0001, 'data': 'D'}{'in_features': [('arm', 0), ('rel3', 0)], 'out_features': [('arm', 1)], 'window': 1024}
	  there is some unstability during training around 800 steps, try gradient clipping?

loaddata.py
import sys
import numpy as np
import torch
import gzip
from torch.utils.data import TensorDataset, Dataset


class CalvinDataset(Dataset):
    """
    Turn the last `instances_per_episode` frames of each episode into instances with `features` from the last `context_length` frames.
    Version 3: This has the same behavior as calvindataset() but is more memory efficient.
    It keeps only the raw data in memory and constructs the context for each instance as needed.
    """
    def __init__(self, prefix='data/debug-training', features=range(1,98), instances_per_episode=1, context_length=64):
        # There are three indices for each frame:
        # 1. Frame id given in the filename (and first column of data), which may be discontinuous and not start from 0.
        # 2. Position in the `data` array. pos2id[2]=1 & id2pos[1]=2.
        # 3. Position in the final dataset, which contains only the annotated subset of data. data_index[3]=2
        data,pos2id,id2pos = loaddata(prefix)
        lang,task2int,int2task = loadlang(prefix)
        data_index = []; target = []; frame = []
        for (start_frame, end_frame, task, annot) in lang:
            taskid = task2int[task]
            for instance_frame in range(end_frame - instances_per_episode + 1, end_frame+1):
                instance_index = id2pos[instance_frame]
                if instance_index - context_length + 1 >= 0:
                    frame.append(instance_frame)
                    target.append(taskid)
                    data_index.append(instance_index)
        self.__dict__.update(locals()) # convert each local variable to self.var

    def __getitem__(self, index):
        i = self.data_index[index]
        x = self.data[np.ix_(range(1+i-self.context_length, 1+i), self.features)]
        inputs = torch.from_numpy(x)
        target = torch.tensor(self.target[index])
        frame = torch.tensor(self.frame[index])
        return (inputs, target, frame)

    def __len__(self):
        return len(self.target)


def calvindataset(prefix='data/debug-training', features=range(1,98), instances_per_episode=32, context_length=1):
    """
    Turn the last `instances_per_episode` frames of each episode into instances with `features` from the last `context_length` frames.
    Version 2: generalizes the three functions calvindataset1,2,3.
    To get equivalent output to the previous version:
    calvindataset1(..., window=32) = calvindataset(..., instances_per_episode=32, context_length=1)
    calvindataset2(..., window=32) = calvindataset(..., instances_per_episode=1, context_length=32)
    ... and except for reordering of features:
    calvindataset3(..., window=32, frames=4) = calvindataset(..., instances_per_episode=32, context_length=4)
    """
    data,pos2id,id2pos = loaddata(prefix)
    lang,task2int,int2task = loadlang(prefix)
    x = []; y = []; idx = []
    for (i,j,task,annot) in lang: # episode starts at i, ends at j, classified as task.
        taskid = task2int[task]
        for k in range(j-instances_per_episode+1, j+1): # we create instances for each of the last n frames of an episode
            p = id2pos[k] - context_length + 1
            if p >= 0:
                idx.append(k)
                y.append(taskid)
                x.append(data[np.ix_(range(p, p+context_length), features)].reshape(-1))
    x = torch.tensor(np.stack(x))
    y = torch.tensor(y)
    idx = torch.tensor(idx)
    return TensorDataset(x,y,idx)


def calvindataset1(prefix='data/debug-training', features=range(1,98), window=32):
    """
    Predict task from a single frame (picked anywhere from the last 32 frames associated with task)
    Version 1: deprecated, please use CalvinDataset(features, instances_per_episode=32, context_length=1)
    """
    data,pos2id,id2pos = loaddata(prefix)
    lang,task2int,int2task = loadlang(prefix)
    p = []
    y = []
    idx = []
    for (i,j,task,annot) in lang: # i,j,k frame ids. p is an index into data.
        taskid = task2int[task]
        for k in range(j-window+1,j+1):
            p.append(id2pos[k])
            y.append(taskid)
            idx.append(k)
    x = torch.tensor(data[np.ix_(p,features)])
    y = torch.tensor(y)
    idx = torch.tensor(idx)
    return TensorDataset(x,y,idx)



def calvindataset2(prefix='data/debug-training', features=range(1,98), window=32):
    """
    Predict annotation from the last 32 frames (out of 64 associated with the annotation).
    Version 1: deprecated, please use CalvinDataset(features, instances_per_episode=1, context_length=32)
    """
    data,pos2id,id2pos = loaddata(prefix)
    lang,task2int,int2task = loadlang(prefix)
    x = []
    y = []
    idx = []
    for (i,j,task,annot) in lang:
        taskid = task2int[task]
        p = id2pos[j]
        x.append(np.ravel(data[np.ix_(range(p-window+1,p+1), features)]))
        y.append(taskid)
        idx.append(j)
    x = torch.tensor(np.stack(x))
    y = torch.tensor(y)
    idx = torch.tensor(idx)
    return TensorDataset(x,y,idx)



def calvindataset3(prefix='data/debug-training', features=range(1,98), window=32, frames=4):
    """
    Predict annotation from 4 successive frames (picked anywhere from the last 32 frames associated with task).
    Version 1: deprecated, please use CalvinDataset(features, instances_per_episode=32, context_length=4)
    """
    data,pos2id,id2pos = loaddata(prefix)
    lang,task2int,int2task = loadlang(prefix)
    p = []
    y = []
    idx = []
    for (i,j,task,annot) in lang:
        taskid = task2int[task]
        for k in range(j-window+1,j+1):
            p.append(id2pos[k])
            y.append(taskid)
            idx.append(k)
    p = np.array(p)
    x = []
    for f in range(0, frames):
        x.append(torch.tensor(data[np.ix_(p-f, features)]))
    x = torch.cat(x, dim=1)
    y = torch.tensor(y)
    idx = torch.tensor(idx)
    return TensorDataset(x,y,idx)


def loaddata(prefix):
    print(f"Loading {prefix}", file=sys.stderr)
    with gzip.open(prefix + '.tsv.gz', 'rt') as f:
        data = np.loadtxt(f, delimiter='\t', dtype='float32') # not using dtype=dtype_data: it makes a 1-D record array instead of 2-D
    with gzip.open(prefix + '-controllers.tsv.gz', 'rt') as f:
        cont = np.loadtxt(f, delimiter='\t', dtype='float32')
        assert np.array_equal(data[:,0], cont[:,0]), 'cont indices do not match'
    with gzip.open(prefix + '-tactile2.tsv.gz', 'rt') as f:
        tact = np.loadtxt(f, delimiter='\t', dtype='float32')
        assert np.array_equal(data[:,0], tact[:,0]), 'tact indices do not match'
    with gzip.open(prefix + '-sdiff.tsv.gz', 'rt') as f:
        sdiff = np.loadtxt(f, delimiter='\t', dtype='float32')
        assert np.array_equal(data[:,0], sdiff[:,0]), 'sdiff indices do not match'
    data = np.concatenate((data, cont[:,1:], tact[:,1:], sdiff[:,1:]), axis=1)
    data = normalize(data)
    pos2id = data[:,0].astype(int)
    id2pos = np.full(1+max(pos2id), -1)
    for (pos, id) in enumerate(pos2id):
        id2pos[id] = pos
    return data, pos2id, id2pos


def normalize(data):
    data[:,32:34] = data[:,32:34] * 10.0  # button and switch
    data[:,21] = data[:,21] * 10.0        # gripper opening
    return data


def loadlang(prefix):
    print(f"Loading {prefix}-lang", file=sys.stderr)
    with gzip.open(prefix + '-lang.tsv.gz', 'rt') as f:
        lang = np.loadtxt(f, delimiter='\t', dtype=dtype_lang)
    lang.sort(order = 'end')
    task2int = {}
    for (i,task) in enumerate(int2task):
        task2int[task] = i
    for task in lang['task']:
        assert task in task2int, 'task not found'
    return lang, task2int, int2task

dtype_lang = [
    ('start', int),
    ('end', int),
    ('task', object),
    ('annot', object)
]


dtype_data = [
    ('idnum', int),  # 00. 
    ('actx', float), # 01.  (tcp (tool center point) position (3): x,y,z in absolute world coordinates)
    ('acty', float), # 02. 
    ('actz', float), # 03. 
    ('acta', float), # 04.  (tcp orientation (3): euler angles a,b,c in absolute world coordinates)
    ('actb', float), # 05. 
    ('actc', float), # 06. 
    ('actg', int),   # 07.  (gripper_action (1): binary close=-1, open=1)
    ('relx', float), # 08.  (tcp position (3): x,y,z in relative world coordinates normalized and clipped to (-1, 1) with scaling factor 50)
    ('rely', float), # 09. 
    ('relz', float), # 10. 
    ('rela', float), # 11.  (tcp orientation (3): euler angles a,b,c in relative world coordinates normalized and clipped to (-1, 1) with scaling factor 20)
    ('relb', float), # 12. 
    ('relc', float), # 13. 
    ('relg', int),   # 14.  (gripper_action (1): binary close=-1, open=1)
    ('tcpx', float), # 15.  (tcp position (3): x,y,z in world coordinates)
    ('tcpy', float), # 16. 
    ('tcpz', float), # 17. 
    ('tcpa', float), # 18.  (tcp orientation (3): euler angles a,b,c in world coordinates)
    ('tcpb', float), # 19. 
    ('tcpc', float), # 20. 
    ('tcpg', float), # 21.  (gripper opening width (1): in meters)
    ('arm1', float), # 22.  (arm_joint_states (7): in rad)
    ('arm2', float), # 23. 
    ('arm3', float), # 24. 
    ('arm4', float), # 25. 
    ('arm5', float), # 26. 
    ('arm6', float), # 27. 
    ('arm7', float), # 28. 
    ('armg', int),   # 29.  (gripper_action (1): binary close = -1, open = 1)
    ('slider', float), # 30.  (1): joint state: range=[-0.002359:0.306696]
    ('drawer', float), # 31.  (1): joint state: range=[-0.002028:0.221432]
    ('button', float), # 32.  (1): joint state: range=[-0.000935:0.033721]
    ('switch', float), # 33.  (1): joint state: range=[-0.004783:0.091777]
    ('lightbulb', int),  # 34.  (1): on=1, off=0
    ('greenlight', int), # 35.  (1): on=1, off=0
    ('redx', float), # 36.  (red block (6): (x, y, z, euler_x, euler_y, euler_z)
    ('redy', float), # 37. 
    ('redz', float), # 38. 
    ('reda', float), # 39. 
    ('redb', float), # 40. 
    ('redc', float), # 41. 
    ('bluex', float), # 42.  (blue block (6): (x, y, z, euler_x, euler_y, euler_z)
    ('bluey', float), # 43. 
    ('bluez', float), # 44. 
    ('bluea', float), # 45. 
    ('blueb', float), # 46. 
    ('bluec', float), # 47. 
    ('pinkx', float), # 48.  (pink block (6): (x, y, z, euler_x, euler_y, euler_z)
    ('pinky', float), # 49. 
    ('pinkz', float), # 50. 
    ('pinka', float), # 51. 
    ('pinkb', float), # 52. 
    ('pinkc', float)  # 53. 
]

dtype_cont = [
    ('idnum', int),      # 00.00
    ('slider.x', float), # 01.54
    ('slider.y', float), # 02.55
    ('slider.z', float), # 03.56
    ('drawer.x', float), # 04.57
    ('drawer.y', float), # 05.58
    ('drawer.z', float), # 06.59
    ('button.x', float), # 07.60
    ('button.y', float), # 08.61
    ('button.z', float), # 09.62
    ('switch.x', float), # 10.63
    ('switch.y', float), # 11.64
    ('switch.z', float)  # 12.65
]

dtype_tact = [
    ('idnum', int),    # 00.00 idnum
    ('tact1d', float), # 01.66 depth_tactile1
    ('tact2d', float), # 02.67 depth_tactile2
    ('tact1r', float), # 03.68 rgb_tactile1_r
    ('tact1g', float), # 04.69 rgb_tactile1_g
    ('tact1b', float), # 05.70 rgb_tactile1_b
    ('tact2r', float), # 06.71 rgb_tactile2_r
    ('tact2g', float), # 07.72 rgb_tactile2_g
    ('tact2b', float)  # 08.73 rgb_tactile2_b
]

dtype_sdiff = [
    ('idnum', int),      # 00.00
    ('slider_diff', float),   # 01.74
    ('drawer_diff', float),   # 02.75
    ('button_diff', float),   # 03.76
    ('switch_diff', float),   # 04.77
    ('lightbulb_diff', int),  # 05.78
    ('greenlight_diff', int), # 06.79
    ('redx_diff', float),     # 07.80
    ('redy_diff', float),     # 08.81
    ('redz_diff', float),     # 09.82
    ('reda_diff', float),     # 10.83
    ('redb_diff', float),     # 11.84
    ('redc_diff', float),     # 12.85
    ('bluex_diff', float),    # 13.86
    ('bluey_diff', float),    # 14.87
    ('bluez_diff', float),    # 15.88
    ('bluea_diff', float),    # 16.89
    ('blueb_diff', float),    # 17.90
    ('bluec_diff', float),    # 18.91
    ('pinkx_diff', float),    # 19.92
    ('pinky_diff', float),    # 20.93
    ('pinkz_diff', float),    # 21.94
    ('pinka_diff', float),    # 22.95
    ('pinkb_diff', float),    # 23.96
    ('pinkc_diff', float)     # 24.97
]    

act_range = range(1,8)
rel_range = range(8,15)
tcp_range = range(15,22)
arm_range = range(22,30)
scene_range = range(30,54)
controller_range = range(54,66)
tactile_range = range(66,74)
sdiff_range = range(74,98)

int2task = [
    'close_drawer',
    'lift_blue_block_drawer',
    'lift_blue_block_slider',
    'lift_blue_block_table',
    'lift_pink_block_drawer',
    'lift_pink_block_slider',
    'lift_pink_block_table',
    'lift_red_block_drawer',
    'lift_red_block_slider',
    'lift_red_block_table',
    'move_slider_left',
    'move_slider_right',
    'open_drawer',
    'place_in_drawer',
    'place_in_slider',
    'push_blue_block_left',
    'push_blue_block_right',
    'push_into_drawer',
    'push_pink_block_left',
    'push_pink_block_right',
    'push_red_block_left',
    'push_red_block_right',
    'rotate_blue_block_left',
    'rotate_blue_block_right',
    'rotate_pink_block_left',
    'rotate_pink_block_right',
    'rotate_red_block_left',
    'rotate_red_block_right',
    'stack_block',
    'turn_off_led',
    'turn_off_lightbulb',
    'turn_on_led',
    'turn_on_lightbulb',
    'unstack_block'
]

